File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def dfs(start, path):
            if start == len(s):
                result.append(path[:])
                return

            for end in range(start + 1, len(s) + 1):
                currSubstr = s[start:end]
                if self.isPalidrome(currSubstr):
                    path.append(currSubstr)
                    dfs(end, path)
                    path.pop()

        result = []
        dfs(0, [])

        return result

    def isPalidrome(self, s):
        i = 0
        j = len(s) - 1

        while i < j:
            if s[i] != s[j]:
                return False
            i += 1
            j -= 1

        return True

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function (s) {
    const dfs = function (start, path) {
        if (start === s.length) {
            result.push([...path]);
            return;
        }

        for (let end = start + 1; end <= s.length; end++) {
            const currSubstr = s.substring(start, end);
            if (isPalindrome(currSubstr)) {
                path.push(currSubstr);
                dfs(end, path);
                path.pop();
            }
        }
    }

    const result = [];
    dfs(0, []);

    return result;
};

const isPalindrome = function (s) {
    let i = 0, j = s.length - 1;

    while (i < j) {
        if (s[i] !== s[j]) return false;
        i++;
        j--;
    }

    return true;
}